### PR TITLE
Configuring search filters for dashboard widgets/core

### PR DESCRIPTION
**Pull request by maxiadlovskii**
## Description
Extend Widget, AggregationWidget, and MessageWidget to class to handle filters

## Motivation and Context
[Feat: #3647](https://github.com/Graylog2/graylog-plugin-enterprise/issues/3647)
[Feat: #3691](https://github.com/Graylog2/graylog-plugin-enterprise/issues/3691)
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

/jenkins-pr-deps https://github.com/Graylog2/graylog-plugin-enterprise/pull/3678

### CODE REVIEW

File: graylog2-web-interface/src/views/logic/widgets/Widget.ts [hunk 1]
```diff
@@ -89,6 +97,7 @@ class Widget {
 
     return this.id === other.id
       && this.filter === other.filter
+      && isDeepEqual(this.filters, other.filters)
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/logic/widgets/Widget.ts:100]: Let's use the same checks here for `fitler` and `filters` like in `Query.ts`, just to be save:
```
|| !((isNullish(this.filter) && isNullish(other.filter)) || isDeepEqual(this.filter, other.filter))
|| !((isNullish(this.filters) && isNullish(other.filters)) || isDeepEqual(this.filters, other.filters))
```

File: graylog2-web-interface/src/views/logic/widgets/MessagesWidget.ts [hunk 2]
```diff
@@ -26,8 +27,9 @@ import type { WidgetState } from './Widget';
 import type { QueryString, TimeRange } from '../queries/Query';
 
 export default class MessagesWidget extends Widget {
-  constructor(id: string, config: any, filter: string | undefined | null, timerange: TimeRange | undefined | null, query: QueryString | undefined | null, streams: Array<string> = []) {
-    super(id, MessagesWidget.type, config, filter, timerange, query, streams);
+  // eslint-disable-next-line default-param-last
+  constructor(id: string, config: any, filter: string | undefined | null, timerange: TimeRange | undefined | null, query: QueryString | undefined | null, streams: Array<string> = [], filters?: FiltersType) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/logic/widgets/MessagesWidget.ts:31]: Afaics we do not rely on the default value for the `streams` parameter, so we can just remove it. This way we do not need to suppress `default-param-last` for the `filters` parameter.

File: graylog2-web-interface/src/views/logic/widgets/Widget.ts [hunk 3]
```diff
@@ -35,14 +38,17 @@ type DeserializesWidgets = {
...
 class Widget {
   _value: WidgetState;
 
   // eslint-disable-next-line @typescript-eslint/no-use-before-define
   static Builder: typeof Builder;
 
-  constructor(id: string, type: string, config: any, filter?: string, timerange?: TimeRange, query?: QueryString, streams: Array<string> = []) {
-    this._value = { id, type, config, filter: filter === null ? undefined : filter, timerange, query, streams };
+  // eslint-disable-next-line default-param-last
+  constructor(id: string, type: string, config: any, filter?: string, timerange?: TimeRange, query?: QueryString, streams: Array<string> = [], filters?: FiltersType) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/logic/widgets/Widget.ts:50]: Afaics we do not rely on the default value for the `streams` parameter, so we can just remove it. This way we do not need to suppress `default-param-last` for the `filters` parameter.

File: graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts [hunk 4]
```diff
@@ -19,49 +19,51 @@ import { Map } from 'immutable';
...
 
 import AggregationWidgetConfig from './AggregationWidgetConfig';
 
 import Widget from '../widgets/Widget';
 
 export default class AggregationWidget extends Widget {
-  constructor(id: string, config: AggregationWidgetConfig, filter?: string, timerange?: TimeRange, query?: QueryString, streams: Array<string> = []) {
-    super(id, AggregationWidget.type, config, filter, timerange, query, streams);
+  // eslint-disable-next-line default-param-last
+  constructor(id: string, config: AggregationWidgetConfig, filter?: string, timerange?: TimeRange, query?: QueryString, streams: Array<string> = [], filters?: FiltersType) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts:30]: Afaics we do not rely on the default value for the `streams` parameter, so we can just remove it. This way we do not need to suppress `default-param-last` for the `filters` parameter.

File: graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts [hunk 5]
```diff
@@ -19,49 +19,51 @@ import { Map } from 'immutable';
...
+    return new AggregationWidget(id, AggregationWidgetConfig.fromJSON(config), filter, timerange, query, streams, filters);
   }
 
   toBuilder() {
-    const { id, config, filter, timerange, query, streams } = this._value;
+    const { id, config, filter, timerange, query, streams, filters } = this._value;
 
-    // eslint-disable-next-line no-use-before-define
-    return new Builder(Map({ id, config, filter, timerange, query, streams }));
+    // eslint-disable-next-line no-use-before-define,@typescript-eslint/no-use-before-define
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts:47]: The `typescript-eslint` hint should replace the previous one:

```suggestion
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
```

File: graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts [hunk 6]
```diff
@@ -19,49 +19,51 @@ import { Map } from 'immutable';
...
+    return new Builder(Map({ id, config, filter, timerange, query, streams, filters }));
   }
 
   static builder() {
-    // eslint-disable-next-line no-use-before-define
+    // eslint-disable-next-line no-use-before-define,@typescript-eslint/no-use-before-define
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/logic/aggregationbuilder/AggregationWidget.ts:52]: Same here:

```suggestion
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
```

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Works well and looks good! Just added some minor comments for small cleanups.

**Pull request merged**